pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/retention/SegmentLineageCleanupTest.java

```diff
@@ -168,15 +175,60 @@ public void testSegmentLineageCleanup()
     segmentsForTable = ControllerTestUtils.getHelixResourceManager().getSegmentsFor(OFFLINE_TABLE_NAME, false);
     Assert.assertEquals(segmentsForTable.size(), 3);
     Assert.assertTrue(Collections.disjoint(segmentsForTable, Arrays.asList("merged_1", "merged_2")));
-    segmentLineage =
-        SegmentLineageAccessHelper
-            .getSegmentLineage(ControllerTestUtils.getHelixResourceManager().getPropertyStore(), OFFLINE_TABLE_NAME);
+    segmentLineage = SegmentLineageAccessHelper
+        .getSegmentLineage(ControllerTestUtils.getHelixResourceManager().getPropertyStore(), OFFLINE_TABLE_NAME);
     Assert.assertEquals(segmentLineage.getLineageEntryIds().size(), 1);
   }
 
+  @Test
+  public void testRefreshTableCleanup()
+      throws InterruptedException {
+    // Create metadata for original segments
+    for (int i = 0; i < 3; i++) {
+      ControllerTestUtils.getHelixResourceManager().addNewSegment(REFRESH_OFFLINE_TABLE_NAME,
+          SegmentMetadataMockUtils.mockSegmentMetadata(REFRESH_OFFLINE_TABLE_NAME, "segment1_" + i), "downloadUrl");
+    }
+
+    // Create metadata for new segments.
+    for (int i = 0; i < 3; i++) {
+      ControllerTestUtils.getHelixResourceManager().addNewSegment(REFRESH_OFFLINE_TABLE_NAME,
+          SegmentMetadataMockUtils.mockSegmentMetadata(REFRESH_OFFLINE_TABLE_NAME, "segment2_" + i), "downloadUrl");
+    }
+
+    Assert.assertEquals(
+        ControllerTestUtils.getHelixResourceManager().getSegmentsFor(REFRESH_OFFLINE_TABLE_NAME, false).size(), 6);
+
+    // Validate the case when the lineage entry state is 'IN_PROGRESS'
+    SegmentLineage segmentLineage = new SegmentLineage(REFRESH_OFFLINE_TABLE_NAME);
+    segmentLineage.addLineageEntry("0", new LineageEntry(Arrays.asList("segment1_0", "segment1_1", "segment1_2"),
+        Arrays.asList("segment2_0", "segment2_1", "segment2_2"), LineageEntryState.IN_PROGRESS,
+        System.currentTimeMillis()));
+    SegmentLineageAccessHelper
+        .writeSegmentLineage(ControllerTestUtils.getHelixResourceManager().getPropertyStore(), segmentLineage, -1);
+    _retentionManager.processTable(REFRESH_OFFLINE_TABLE_NAME);
+    try {
+      waitForSegmentsToDelete(REFRESH_OFFLINE_TABLE_NAME, 3, 1000L);
```

Comment:
You can skip the 3rd parameter.

Reply:
This will directly add 10 sec to the overall testing. I thought this may be too much. I prefer to keep it.

Reply:
Setting the 3rd param to 1000L means the validation will only be done at most 2 times ( probably just 1 time in limited resources). The default 10 seconds doesn't mean it always wait for 10 seconds. But I'm fine with it.